### PR TITLE
CONIKS, hstar2 and SMT Writer performance improvements

### PR DESCRIPTION
**Pull request by Martin2112**
Bunch of performance improvements to CONIKS hasher, hstar2 and Sparse Merkle Tree Writer. Only affects Map trees so should be fairly safe.

Benchmarking shows ~20% speed increase and ~3% RAM allocation reduction for the combined changes. Some of the speedup comes from switching to pointers for `HStar2LeafHash`.

### Checklist

- [ ] I have updated the [CHANGELOG](CHANGELOG.md).
  - Adjust the draft version number according to [semantic versioning](https://semver.org/) rules.
- [ ] I have updated [documentation](docs/) accordingly.

### CODE REVIEW

File: merkle/coniks/coniks.go [hunk 1]
```diff
@@ -106,25 +113,22 @@ func (m *hasher) BitLen() int {
...
 var leftmask = [8]byte{0xFF, 0x80, 0xC0, 0xE0, 0xF0, 0xF8, 0xFC, 0xFE}
 
-// maskIndex returns index with only the left depth bits set.
-// index must be of size m.Size() and 0 <= depth <= m.BitLen().
-// e.g.
-func (m *hasher) maskIndex(index []byte, depth int) []byte {
+// maskIndex writes the masked value directly to a Buffer (which never
+// returns an error on writes). This avoids the need to allocate space for and
+// copy a value that will then be discarded immediately.
+func (m *hasher) maskIndex(b *bytes.Buffer, index []byte, depth int) {
```
Comment (AlCutter) [inline merkle/coniks/coniks.go]: rename to `writeMaskedIndex`?

File: merkle/coniks/coniks.go [hunk 2]
```diff
@@ -124,11 +125,19 @@ func (m *hasher) maskIndex(b *bytes.Buffer, index []byte, depth int) {
 		panic(fmt.Sprintf("depth: %d, want <= %d && > 0", got, want))
 	}
 
+	prevLen := b.Len()
 	if depth > 0 {
 		// Copy the first depthBytes.
 		depthBytes := (depth + 7) >> 3
-		b.Write(index[:depthBytes])
+		if depth > 1 {
```
Comment (AlCutter) [inline merkle/coniks/coniks.go]: Should we add a test for this case?

Comment (Martin2112) [reply merkle/coniks/coniks.go]: I think it's covered by existing tests but I'll go back and make sure.

File: merkle/coniks/coniks.go [hunk 3]
```diff
@@ -60,11 +62,13 @@ func (m *hasher) EmptyRoot() []byte {
 func (m *hasher) HashEmpty(treeID int64, index []byte, height int) []byte {
 	depth := m.BitLen() - height
 
+	buf := new(bytes.Buffer)
```
Comment (gdbelvin) [inline merkle/coniks/coniks.go]: Do you think it would be worth preallocating the buffer to avoid any runtime resizing?

Comment (Martin2112) [reply merkle/coniks/coniks.go]: Probably only if it resizes more than once, which is a possibility. I can test that with a benchmark.

File: merkle/coniks/coniks.go (hunk 2 above)
Comment (Martin2112) [reply merkle/coniks/coniks.go]: This was incorrect, should have been `depthBytes`. However for depthBytes to be zero depth must be zero so it worked anyway by luck.

File: merkle/coniks/coniks.go (hunk 1 above)
Comment (Martin2112) [reply merkle/coniks/coniks.go]: Done.

File: merkle/sparse_merkle_tree.go [hunk 4]
```diff
@@ -257,7 +254,7 @@ func (s *subtreeWriter) buildSubtree(ctx context.Context, queueSize int) {
 		}
 		span.End()
 
-		hsCtx, span := spanFor(ctx, "buildSubtree.runTX.htar2")
+		hsCtx, span := spanFor(ctx, "buildSubtree.runTX.hstar2")
```
Comment (AlCutter) [inline merkle/sparse_merkle_tree.go]: Aha, all this extra `tar` must've been slowing it down! :)

File: merkle/coniks/coniks.go (hunk 2 above)
Comment (Martin2112) [reply merkle/coniks/coniks.go]: Additional tests added as a crosscheck against bit twiddling on a big.Int.

File: merkle/coniks/coniks.go (hunk 3 above)
Comment (Martin2112) [reply merkle/coniks/coniks.go]: Seems about 5% faster. Might be worth it though the time in this function is probably way less than in the hashing.

**Pull request merged**